### PR TITLE
Write some call stack tests

### LINKED ISSUE #650: call frames [mochitests]
Sub task of the All the mochitests! #634 issue

browser_dbg_stack-01.js
browser_dbg_stack-02.js
browser_dbg_stack-03.js
browser_dbg_stack-04.js
browser_dbg_stack-05.js
browser_dbg_stack-06.js
browser_dbg_stack-07.js

[DXR search](https://dxr.mozilla.org/mozilla-central/search?q=path%3Adevtools%2Fclient%2Fdebugger%2Ftest%2Fmochitest%2F+file%3A*dbg_stack-%5B0-9%5D*&redirect=false)

### PR DESCRIPTION
**Pull request by jasonLaster**
Fixes #650

### CODE REVIEW

File: public/js/test/mochitest/examples/code_script-switching-02.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+function secondCall() {
+  // This comment is useful: ☺
+  debugger;
+  function foo() {}
+  if (x) {
+    foo();
+  }
+}
+
+var x = true;
```
Comment (jasonLaster) [inline public/js/test/mochitest/examples/code_script-switching-02.js]: I pulled in this example from the previous tests, but i'd be happy to use `code-simple` and add a breakpoint to pause

File: public/js/test/mochitest/examples/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+### Test Examples
+
+##### Pages
+* **doc_script-switching-01** - includes two scripts that reference each other. The second function has a debugger.
+* **doc-scripts** - includes three sources, a long source and two sources that reference each other.
```
Comment (jasonLaster) [inline public/js/test/mochitest/examples/README.md]: I'd like to start documenting our test examples from the start from the beginning. It is really easy to add more and more examples when there is a perfectly good one already.

File: public/js/test/mochitest/head.js [hunk 3]
```diff
@@ -204,3 +221,27 @@ function isVisibleWithin(outerEl, innerEl) {
   return innerRect.top > outerRect.top &&
     innerRect.bottom < outerRect.bottom;
 }
+
+
+const selectors = {
+  callStackArrow: ".call-stack-pane .arrow",
+  frame: (index) => `.frames ul li:nth-child(${index})`
+}
```
Comment (jlongster) [inline public/js/test/mochitest/head.js]: This is awesome. I was thinking about something like this as well instead of having a ton of `get*` functions.

**Comment by jlongster** [comment]
The `editor-mode` test timed out but that has been fixed on master. The problem was that that script it depended on was being GCed because it didn't introduce any global variables, so it was racy. You can ignore that failure or rebase.

File: public/js/test/mochitest/browser_dbg_call-stack.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+  invokeInTab("firstCall");
+  yield waitForPaused(dbg);
+
+  ok(isFrameSelected(dbg, 1, "secondCall"), "the first frame is selected");
+
+  findFrame(dbg, 2).click();
+  ok(isFrameSelected(dbg, 2, "firstCall"), "the second frame is selected");
+
+  ok(true);
```
Comment (jlongster) [inline public/js/test/mochitest/browser_dbg_call-stack.js]: You shouldn't need this if you have other checks in the test; this was in the stub because it needed at least one `ok`

File: public/js/test/mochitest/head.js (hunk 3 above)
Comment (jasonLaster) [inline public/js/test/mochitest/head.js]: yeah - i've used this pattern a lot in other projects and is both concise and descriptive. I was struggling with differentiating UI and state methods like `getFrame`and I think this helps with that as well.

File: public/js/test/mochitest/examples/README.md (hunk 2 above)
Comment (jlongster) [inline public/js/test/mochitest/examples/README.md]: Any chance we could document them at the top of the HTML files instead? I think it'll be easier to keep up-to-date. Usually we add more edge cases to existing test files over time so they'll probably cover more than these simple descriptions. We should just r- PRs that add new examples unless it's really necessary.

If this really clarifies things for you though, I'm willing to try it.

Comment (jlongster) [inline public/js/test/mochitest/examples/README.md]: Also can you rename the new test file with the `doc-` prefix instead of `doc_`? Trying to avoid mixing underscores and dashes (but `browser_` is required by mochitest)

File: public/js/test/mochitest/examples/code_script-switching-02.js (hunk 1 above)
Comment (jlongster) [inline public/js/test/mochitest/examples/code_script-switching-02.js]: I don't mind having several test cases available that do a few different things. I think this is a good one to have too.

File: public/js/test/mochitest/head.js [hunk 5]
```diff
@@ -204,3 +221,27 @@ function isVisibleWithin(outerEl, innerEl) {
...
+  return dbg.win.document.querySelector(selector);
+}
+
+function toggleCallStack(dbg) {
+  return findElement(dbg, "callStackArrow").click()
+}
+
+Services.prefs.setBoolPref("devtools.debugger.new-debugger-frontend", true);
+registerCleanupFunction(() => {
+  Services.prefs.setBoolPref("devtools.debugger.new-debugger-frontend", false);
```
Comment (jlongster) [inline public/js/test/mochitest/head.js]: I think you've duplicated this. It's already at the top of `head.js`, so we don't need this here.

**Comment by jlongster** [comment]
👍 Looks great to me! A few notes, but I like it.

File: public/js/test/mochitest/head.js (hunk 3 above)
Comment (jlongster) [inline public/js/test/mochitest/head.js]: Yep, that was my concern too, that we'd be replicating selectors and it'd get confusing... this pattern is great though. I don't think we should be too stringent about keeping selectors out of tests, but common ones should certainly be in here.

File: public/js/test/mochitest/examples/README.md (hunk 2 above)
Comment (jasonLaster) [inline public/js/test/mochitest/examples/README.md]: ha - yeah the `-` and `_` bothered me too. I left it in as i was just removing the code switching example.

I'd like to try keeping the readme up to date if that's okay. With tests it helps me to have a global spec i can see after each run to understand what's being covered. We don't really have that for the examples and it might be nice to skim the list to have a sense. I hate reading one example and then looking at another and trying to keep it all in my head. I can see this getting out of date if we don't remember it, so i'd be happy to abandon the experiment in a week if it doesn't work for us

File: public/js/test/mochitest/browser_dbg_call-stack.js (hunk 4 above)
Comment (jasonLaster) [inline public/js/test/mochitest/browser_dbg_call-stack.js]: cool

File: public/js/test/mochitest/head.js (hunk 5 above)
Comment (jasonLaster) [inline public/js/test/mochitest/head.js]: oh nice

File: public/js/test/mochitest/examples/code_script-switching-02.js (hunk 1 above)
Comment (jasonLaster) [inline public/js/test/mochitest/examples/code_script-switching-02.js]: cool

File: public/js/test/mochitest/examples/README.md (hunk 2 above)
Comment (jlongster) [inline public/js/test/mochitest/examples/README.md]: Ok, we can try it. I agree that we should avoid duplicating test cases. In the old tests, most of them use just a few of the examples. The two that we have right now will probably cover a lot of them. Of course, there are plenty of edge cases that will need new examples for specific things. Maybe in the readme we could make a not about the "common" test cases that you should probably use unless you know what you're doing.

**Comment by jasonLaster** [comment]
The only tests that failed were cypress
